src/verify_cert.rs

```diff
@@ -481,39 +466,131 @@ where
     Err(error.into())
 }
 
-/// A node in a [`Cert`] path, represented as a linked list from trust anchor to end-entity.
+/// A path for consideration in path building.
+///
+/// This represents a partial path because it does not yet contain the trust anchor. It stores
+/// the end-entity certificates, and an array of intermediate certificates.
+pub(crate) struct PartialPath<'a> {
+    end_entity: &'a EndEntityCert<'a>,
+    /// Intermediate certificates, in order from end-entity to trust anchor.
+    ///
+    /// Invariant: all values below `used` are `Some`.
+    intermediates: [Option<Cert<'a>>; MAX_SUB_CA_COUNT],
+    /// The number of `Some` values in `intermediates`.
+    ///
+    /// The next `Cert` passed to `push()` will be placed at `intermediates[used]`.
+    /// If this value is 0, the path contains only the end-entity certificate.
+    used: usize,
+}
+
+impl<'a> PartialPath<'a> {
+    pub(crate) fn new(end_entity: &'a EndEntityCert<'a>) -> Self {
+        Self {
+            end_entity,
+            intermediates: Default::default(),
+            used: 0,
+        }
+    }
+
+    pub(crate) fn push(&mut self, cert: Cert<'a>) -> Result<(), ControlFlow<Error, Error>> {
+        if self.used >= MAX_SUB_CA_COUNT {
+            return Err(Error::MaximumPathDepthExceeded.into());
+        }
+
+        self.intermediates[self.used] = Some(cert);
+        self.used += 1;
+        Ok(())
+    }
+
+    fn pop(&mut self) {
+        debug_assert!(self.used > 0);
+        if self.used == 0 {
+            return;
+        }
+
+        self.used -= 1;
+        self.intermediates[self.used] = None;
+    }
+
+    pub(crate) fn node(&self) -> PathNode<'_> {
+        PathNode {
+            path: self,
+            index: self.used,
+            cert: self.head(),
+        }
+    }
+
+    /// Current head of the path.
+    pub(crate) fn head(&self) -> &Cert<'a> {
+        self.get(self.used)
+    }
+
+    /// Get the certificate at index `idx` in the path.
+    ///
+    // `idx` must be in the range `0..=self.used`; `idx` 0 thus yields the `end_entity`,
+    // while subsequent indexes yield the intermediate at `self.intermediates[idx - 1]`.
+    fn get(&self, idx: usize) -> &Cert<'a> {
+        match idx {
+            0 => self.end_entity,
+            _ => self.intermediates[idx - 1].as_ref().unwrap(),
+        }
+    }
+
+    pub(crate) fn role(&self) -> Role {
+        match self.used {
+            0 => Role::EndEntity,
+            _ => Role::Issuer,
+        }
+    }
```

Comment:
It feels a little bit confusing to have this `role` implementation and also the `PathNode::role` implementation. Couldn't a consumer use `node().role()` instead of this?